electronics_lib/Passives.py

```diff
@@ -313,6 +314,120 @@ def derated_capacitance(row: Dict[str, Any]) -> Tuple[float, float]:
+
+    self.generator(self.select_capacitor_no_prod_table, self.capacitance, self.voltage, self.single_nominal_capacitance,
+                   self.part_spec, self.footprint_spec)
+
+    # Output values
+    self.selected_capacitance = self.Parameter(RangeExpr())
+    self.selected_derated_capacitance = self.Parameter(RangeExpr())
+    self.selected_voltage_rating = self.Parameter(RangeExpr())
+
+  DERATE_VOLTCO = {  # in terms of %capacitance / V over 3.6
+    #  'Capacitor_SMD:C_0603_1608Metric'  # not supported, should not generate below 1uF
+    'Capacitor_SMD:C_0805_2012Metric': 0.08,
+    'Capacitor_SMD:C_1206_3216Metric': 0.04,
+  }
+
+  PACKAGES = [
+    'Capacitor_SMD:C_0603_1608Metric',
+    'Capacitor_SMD:C_0805_2012Metric',
+    'Capacitor_SMD:C_1206_3216Metric'
+  ]
+
+  def select_capacitor_no_prod_table(self, capacitance: RangeVal, voltage: RangeVal,
+                                     single_nominal_capacitance: RangeVal,
+                                     part_spec: str, footprint_spec: str) -> None:
+    # test with
+    # python -m unittest electronics_lib.test_capacitor
+
+    # capacitance: user-specified capacitance
+    # single nominal capacitance: no single cap with requested capacitance, must generate multiple parallel caps
+    # nominal capacitance: selected part's capacitance
+
+    value = choose_preferred_number(capacitance, 0, self.E24_SERIES_ZIGZAG, 2)
+    single_cap_max = single_nominal_capacitance[1]
+
+    # enforce minimum package size by capacitance
+    #   if > 1.1 uF, only use 0805 or 1206
+    #   if > 11 uF, only use 1206
+    # derate capacitance
+    #   if voltage < 3.6, don't derate
+    #   if capacitance < 1uF, don't derate
+    #   if footprint not in self.DERATE_VOLTCO, can't derate, generate obviously bogus and ignored value
+    #   else derate by returning (
+    #         row['capacitance'][0] * (1 - voltco * (voltage[1] - 3.6)),
+    #         row['capacitance'][1] * (1 - voltco * (voltage[0] - 3.6))
+    #       )
+
+    if (capacitance[0] + capacitance[1]) / 2 > 11e-6:
+      self.min_package_size = self.PACKAGES[2]
+    elif (capacitance[0] + capacitance[1]) / 2 > 1.1e-6:
+      self.min_package_size = self.PACKAGES[1]
+    else:
+      self.min_package_size = self.PACKAGES[0]
+
+    if self.get(self.footprint_spec) == "":
+      self.footprint_spec = self.min_package_size
+
+    def derated_capacitance() -> Tuple[float, float]:
+      if ((voltage[0] + voltage[1])/2 < 3.6) or ((capacitance[0] + capacitance[1]) / 2 <= 1e-6):
+        return capacitance
+      elif self.min_package_size not in self.DERATE_VOLTCO:
+        return (0, 0)
+      else:
+        voltco = self.DERATE_VOLTCO[self.min_package_size]
+        return (
+          capacitance[0] * (1 - voltco * (voltage[1] - 3.6)),
+          capacitance[1] * (1 - voltco * (voltage[0] - 3.6))
+        )
+
+    self.derated_capacitance = derated_capacitance()
```

Comment:
One thing to keep in mind: the algorithm here is going to look different than as implemented with the part table, since what that did is calculate the derated capacitance given the actual part (with set package and capacitance). Here, you're trying to calculate the package and nominal (label) capacitance given a capacitance spec (which we assume to be post-derating) - so it's kind of a reverse algorithm. Note that in this case, you may need to iterate up the package sizes - as you derate from a smaller package size, it may push up into a larger package size which has a less aggressive derating curve.